crates/bevy_ecs/src/ptr.rs

```diff
@@ -0,0 +1,105 @@
+use std::{cell::UnsafeCell, marker::PhantomData, mem::MaybeUninit, ptr::NonNull};
+
+/// Type-erased pointer into memory. Guaranteed to be correctly aligned, non-null and safe to read for a particular type.
+#[derive(Copy, Clone)]
+pub struct Ptr<'a>(NonNull<u8>, PhantomData<&'a u8>);
+
+/// Type-erased pointer into memory. Guaranteed to be correctly aligned, non-null and safe to modify for a particular type.
+pub struct PtrMut<'a>(NonNull<u8>, PhantomData<&'a mut u8>);
+
+/// Type-erased pointer into memory. Guaranteed to be correctly aligned, non-null and safe to move out of for a particular type.
+pub struct OwningPtr<'a>(NonNull<u8>, PhantomData<&'a mut u8>);
+
+macro_rules! impl_ptr {
+    ($ptr:ident) => {
+        impl $ptr<'_> {
+            /// # Safety
+            /// the offset cannot make the existing ptr null, or take it out of bounds for it's allocation.
+            pub unsafe fn offset(self, count: isize) -> Self {
+                Self(
+                    NonNull::new_unchecked(self.0.as_ptr().offset(count)),
+                    PhantomData,
+                )
+            }
+
+            /// # Safety
+            /// the offset cannot make the existing ptr null, or take it out of bounds for it's allocation.
+            pub unsafe fn add(self, count: usize) -> Self {
+                Self(
+                    NonNull::new_unchecked(self.0.as_ptr().add(count)),
+                    PhantomData,
+                )
+            }
+
+            /// # Safety
+            /// the lifetime for the returned item must not exceed the lifetime `inner` is valid for
+            pub unsafe fn new(inner: NonNull<u8>) -> Self {
```

Comment:
This seems extremely dangerous to use due to inferred lifetimes: this is very easy to mess up and will result in UB lifetime extensions if you do.

Reply:
This problem is repeated elsewhere, e.g. on `OwningPointer::make`.

Reply:
There is nothing that can be done about this, its the same problem as dereferencing a raw pointer creating an unbounded lifetime which has to happen at some point anyway to turn it into a reference.